src/kv/lazy-store.ts

```diff
@@ -0,0 +1,202 @@
+    const release = await this._rwLock.read();
+    return new ReadImpl(this._cache, this._base, release);
+  }
+
+  async withRead<R>(fn: (read: Read) => R | Promise<R>): Promise<R> {
+    const read = await this.read();
+    try {
+      return await fn(read);
+    } finally {
+      read.release();
+    }
+  }
+
+  async write(): Promise<Write> {
+    const release = await this._rwLock.write();
+    return new WriteImpl(
+      this._cache,
+      new ReadImpl(this._cache, this._base, release),
+    );
+  }
+
+  async withWrite<R>(fn: (write: Write) => R | Promise<R>): Promise<R> {
+    const write = await this.write();
+    try {
+      return await fn(write);
+    } finally {
+      write.release();
+    }
+  }
+
+  close(): Promise<void> {
+    return this._base.close();
+  }
+
+  get closed(): boolean {
+    return this._base.closed;
+  }
+}
+
+class ReadImpl implements Read {
+  private readonly _cache: Cache;
+  private readonly _release: () => void;
+  private readonly _base: Store;
+  private _baseRead: Read | undefined;
+  private _closed = false;
+
+  constructor(cache: Cache, base: Store, release: () => void) {
+    this._cache = cache;
+    this._base = base;
+    this._release = release;
+  }
+
+  release() {
+    if (this._baseRead) {
+      this._baseRead.release();
+    }
+    this._release();
+    this._closed = true;
+  }
+
+  get closed(): boolean {
+    return this._closed;
+  }
+
+  async has(key: string): Promise<boolean> {
+    return this.get(key) !== undefined;
+  }
+
+  async get(key: string): Promise<Value | undefined> {
+    let v = this._cache.get(key);
+    if (v === undefined) {
+      v = await (await this._getBaseRead()).get(key);
+      if (v !== undefined) {
+        this._cache.set(key, v);
+      }
+    }
+    return v;
+  }
+
+  private async _getBaseRead(): Promise<Read> {
+    if (!this._baseRead) {
+      this._baseRead = await this._base.read();
+    }
+    return this._baseRead;
+  }
+}
+
+class WriteImpl extends WriteImplBase implements Write {
+  private readonly _cache: Cache;
+
+  constructor(cache: Cache, read: Read) {
+    super(read);
+    this._cache = cache;
+  }
+
+  async commit(): Promise<void> {
+    // HOT. Do not allocate entry tuple and destructure.
+    const ps: Promise<void>[] = [];
+    this._pending.forEach((value, key) => {
+      if (value === deleteSentinel) {
+        this._cache.delete(key);
+      } else {
+        this._cache.set(key, value);
+      }
+    });
+    await Promise.all(ps);
+    this._pending.clear();
+    this.release();
+  }
+}
+
+type CacheEntry = {
+  value: Value;
+  size: number;
+};
+
+class Cache {
+  private readonly _cacheSizeLimit: number;
+  private readonly _shouldBePinned: (key: string) => boolean;
+  private readonly _pinned = new Map<string, Value>();
+  private readonly _cacheEntries = new Map<string, CacheEntry>();
+  private _size = 0;
+
+  constructor(
+    cacheSizeLimit: number,
+    shouldBePinned: (key: string) => boolean,
```

Comment:
Is the idea that `shouldBePinned` will return true if the key is a head? Or if the key refers to a temp hash? Or both? Some comments here would be helpful (realize this is draft and you and Erik probably talked about this).

Reply:
Exactly, shouldBePinned will be true for head keys and chunk keys with temp hashes. 

The tricky thing here is the abstraction layering.  With the current layering, this layer of abstraction should not know about chunks and heads etc... its just a key value store.  That is why I opted for a `shouldBePinned` function to be passed in rather than just explicitly checking for head keys/temp hashes.  For the same reason I'm not sure if we should put comments here about chunks keys/temp hashes. 

In this case though keeping the layer "clean" makes the design less clear.

Reply:
I think a comment given our intended use case is helpful even when the solution is more generic